lib/proxy/client.go

```diff
@@ -0,0 +1,563 @@
+	*grpc.ClientConn
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     *sync.WaitGroup
+
+	id   string
+	addr string
+}
+
+// Client is a peer proxy service client using grpc and tls.
+type Client struct {
+	sync.RWMutex
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	config   ClientConfig
+	conns    map[string]*clientConn
+	metrics  *clientMetrics
+	reporter *reporter
+}
+
+// NewClient creats a new peer proxy client.
+func NewClient(config ClientConfig) (*Client, error) {
+	err := config.checkAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	metrics, err := newClientMetrics()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reporter := newReporter(metrics)
+
+	closeContext, cancel := context.WithCancel(config.Context)
+
+	c := &Client{
+		config:   config,
+		ctx:      closeContext,
+		cancel:   cancel,
+		conns:    make(map[string]*clientConn),
+		metrics:  metrics,
+		reporter: reporter,
+	}
+
+	go c.monitor()
+
+	if c.config.sync != nil {
+		go c.config.sync()
+	} else {
+		go c.sync()
+	}
+
+	return c, nil
+}
+
+// monitor monitors the status of peer proxy grpc connections.
+func (c *Client) monitor() {
+	ticker := c.config.Clock.NewTicker(defaults.ResyncInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.Chan():
+			c.RLock()
+			c.reporter.resetConnections()
+			for _, conn := range c.conns {
+				switch conn.GetState() {
+				case connectivity.Idle:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Idle.String())
+				case connectivity.Connecting:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Connecting.String())
+				case connectivity.Ready:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Ready.String())
+				case connectivity.TransientFailure:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.TransientFailure.String())
+				case connectivity.Shutdown:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Shutdown.String())
+				}
+			}
+			c.RUnlock()
+		}
+	}
+}
+
+// sync runs the peer proxy watcher functionality.
+func (c *Client) sync() {
+	proxyWatcher, err := services.NewProxyWatcher(c.ctx, services.ProxyWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.Component(teleport.ComponentProxyPeer),
+			Client:    c.config.AccessPoint,
+			Log:       c.config.Log,
+		},
+	})
+	if err != nil {
+		c.config.Log.Errorf("Error initializing proxy peer watcher: %+v.", err)
+		return
+	}
+	defer proxyWatcher.Close()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			c.config.Log.Debug("Stopping peer proxy sync.")
+			return
+		case proxies := <-proxyWatcher.ProxiesC:
+			if err := c.updateConnections(proxies); err != nil {
+				c.config.Log.Errorf("Error syncing peer proxies: %+v.", err)
+			}
+		}
+	}
+}
+
+func (c *Client) updateConnections(proxies []types.Server) error {
+	c.RLock()
+	var errs []error
```

Comment:
nit: This is 40 lines north of where it is first used, can it be moved a little closer?